test/lsp.py

```diff
@@ -1296,6 +1308,40 @@ def user_interaction_failed_autoupdate(self, test, sub_dir):
 
     # {{{ actual tests
 
+    def test_analyze_all_project_files(self, solc: JsonRpcProcess) -> None:
```

Comment:
Can you add a more complex test, that would also have some include dirs, and some files that should **not** be detected?

Preferably also symlinks but I guess it would not be a good idea to store these in the repo so you'd have to do that in a temp dir... Does the current testing framework support that?

Reply:
We should also have a test for the case of `analyze_all_files_in_project=False`. Overall, I think this is a bit undertested given how many corner cases there could be.

Reply:
> We should also have a test for the case of `analyze_all_files_in_project=False`. Overall, I think this is a bit undertested given how many corner cases there could be.

@cameel This is confusing. This is the default argument for the `setup_lsp()` And most of the test would actually fail if that would not be supported. It's used in all other tests but those that do explicitly state `analyze_all_files_in_project=True`. If that would not be working, tons of failures would arise when running lsp.py. What exactly is "a bit undertested" here?

> Can you add a more complex test, that would also have some include dirs, and some files that should not be detected?

You are asking for a more complex project and see if it'll analyze them all as expected. What files would you expect it should not load, such as .txt files I assume? 

> Preferably also symlinks but I guess it would not be a good idea to store these in the repo so you'd have to do that in a temp dir... Does the current testing framework support that?

It's supported if it's coded. So for now there is no framework for automatically spinning up some temp dir and populating it with test files as there was no need for that. Symlinks are actually followed (now), and I think it's not really a problem to add a symlink to the repo, it should all work/run just fine. The only concern I would have is how that works on Windows (NTFS), which does have symlinks, but that's via .lnk files, not sure how well that works with recursive directory traversal at all and if we're actually testing our impl for `boost::filesystem::recursive_directory_iterator` here.

Reply:
Call Summary:

- [ ] `analyze-all-files-in-project` = None
- [ ] nested directories
- [ ] another more complex test case containing nested dirs, with symlink(maybe?) & files importing others in its tree.
- [ ] add test that also has include paths

Reply:
> The only concern I would have is how that works on Windows (NTFS), which does have symlinks, but that's via .lnk files

That's what I meant. I know git supports adding symlinks to the repo but they would not be recreated as actual symlinks on Windows.

Also, `.lnk` files are not the equivalent of symlinks. Windows does have actual symlinks, though they're actually a pain to work with. The semantics are different in corner cases. For example there are separate symlink types for files and dirs and if you create the wrong type, you get very weird results.

Reply:
@cameel so let's not support symlinks at all. If it works, it works (like on Unix), but if it isn't, that's simply then not supported. I do not see a major use of this.

Reply:
I'm fine with not testing them due to difficulties of setting that up, but I think they should still be supported - simply because the compiler does support them. Otherwise LSP won't be able to handle some project layouts that the compiler has no problems with.

One of the use cases where some people could have plausibly used symlinks is to work around the `@` imports on versions that did not have `--include-path`. This could have been done via remappings but links were a viable option as well.